include/torch-mlir/Dialect/Torch/IR/TorchOps.td

```diff
@@ -1337,4 +1338,67 @@ def Torch_DtypeCalculateYieldDtypesOp : Torch_Op<"dtype.calculate.yield.dtypes",
   let hasVerifier = 1;
 }
 
+//===----------------------------------------------------------------------===//
+// Symbolic shape modeling ops for TorchDynamo frontend.
+//===----------------------------------------------------------------------===//
+
+def Torch_SymbolicIntOp : Torch_Op<"symbolic_int", [Pure]> {
+  let summary = "Symbolic int representing a dynamic dimension";
+  let description = [{
+    The `torch.symbolic_int` operation captures a dynamic dimension on the
+    global function arguments as exported by TorchDynamo (torch.export).
+    It associates the shape symbols (i.e. "s0", "s1") with the
+    global SSA values (i.e. `%0`, `%1`) that is then referenced
+    to bind shapes on op results.
+
+    Additionally, the operation annotates `min_val` and `max_val` attributes
+    denoting the range constraints for the dynamic dimension. This may be
+    useful for modeling runtime shape guards, or compile-time optimizations
+    based on the shape bounds (min, opt, max) on results of ops / regions.
+
+    Example:
+    ```
+    %0 = torch.symbolic_int "s0" {min_val = 5, max_val = 10} : !torch.int
+    %1 = torch.symbolic_int "s1" {min_val = 2, max_val = 20} : !torch.int
+    ```
+  }];
+  let arguments = (ins
+    StrAttr:$symbol_name,
+    I64Attr:$min_val,
+    I64Attr:$max_val
+  );
+  let results = (outs
+    Torch_IntType:$result
+  );
+  let assemblyFormat = [{
+    $symbol_name ` ` `{` `min_val` `=` $min_val `,` `max_val` `=` $max_val `}` attr-dict `:` type($result)
+  }];
+}
+
+def Torch_BindSymbolicShapeOp : Torch_Op<"bind_symbolic_shape", []> {
+  let summary = "Binds shape expressions to tensors using an affine map indexed by shape symbols";
+  let description = [{
+    The `torch.bind_symbolic_shape` operation binds shape expressions
+    useful to compute the dynamic dimensions of a tensor. It takes a
+    variadic of SSA symbols that map 1:1 to the local symbols declared
+    in the affine map. The affine map contains a list of affine shape
+    expressions for each dim where the terminals are from the declared
+    symbols.
+
+    Example:
+    ```
+    torch.bind_symbolic_shape %arg0, [%0, %1], affine_map<()[s0, s1] -> (s0, s1, 3)> : !torch.vtensor<[?,?,3],f32>
+    torch.bind_symbolic_shape %out0, [%0, %1, %2], affine_map<()[s0, s1, s2] -> (s0, s1 * 2 + s2, 3)> : !torch.vtensor<[?,?,3],f32>
+    ```
+  }];
+  let arguments = (ins
+    Torch_ValueTensorType:$operand,
+    Variadic<Torch_IntType>:$shape_symbols,
```

Comment:
What happens if the shape symbols are not purely generated from torch.symbolic_int or scalar literals ? How about if they are values read from a (not necessarily const) tensor ? 

While I cannot yet find a relevant example in PyTorch, this is seen during the decomposition of tf/tfl.batch_to_space_nd and similar ops. 

This does fall within the class of data dependent dynamism, and it's ok to say this PR itself doesn't go so far as as to address it and that's fine - I'm just pattern matching against several dynamic cases of interest I'm looking at too.

Reply:
I don't think Torch allows that and the cases where it comes up are special cased. Something to watch but I don't think needs to be modeled here.

Reply:
Marking as resolved for now, feel free to reopen if you disagree.